src/simsopt/geo/curveperturbed.py

```diff
@@ -0,0 +1,148 @@
+    functions.
+    Also has the ability to sample the derivatives of the function.
+    """
+
+    def __init__(self, points, sigma, length_scale, n_derivs=1):
+        self.points = points
+        xs = self.points
+        n = len(xs)
+        self.n_derivs = n_derivs
+        cov_mat = np.zeros((n*(n_derivs+1), n*(n_derivs+1)))
+
+        def kernel(x, y):
+            return sum((sigma**2)*exp(-(x-y+i)**2/(length_scale**2)) for i in range(-5, 6))
+
+        XX, YY = np.meshgrid(xs, xs, indexing='ij')
+        x = Symbol("x")
+        y = Symbol("y")
+        f = kernel(x, y)
+        for ii in range(n_derivs+1):
+            for jj in range(n_derivs+1):
+                if ii + jj == 0:
+                    lam = lambdify((x, y), f, "numpy")
+                else:
+                    lam = lambdify((x, y), f.diff(*(ii * [x] + jj * [y])), "numpy")
+                cov_mat[(ii*n):((ii+1)*n), (jj*n):((jj+1)*n)] = lam(XX, YY)
+
+        from scipy.linalg import sqrtm
+        self.L = np.real(sqrtm(cov_mat))
+
+    def draw_sample(self, randomgen=None):
+        n = len(self.points)
+        n_derivs = self.n_derivs
+        if randomgen is None:
+            randomgen = np.random
+        z = randomgen.standard_normal(size=(n*(n_derivs+1), 3))
+        curve_and_derivs = self.L@z
+        return [curve_and_derivs[(i*n):((i+1)*n), :] for i in range(n_derivs+1)]
+
+
+class PerturbationSample():
+
+    def __init__(self, sampler, randomgen=None):
+        self.sampler = sampler
+        self.randomgen = randomgen
+        self.resample()
+
+    def resample(self):
+        self.__sample = self.sampler.draw_sample(self.randomgen)
+
+    def __getitem__(self, deriv):
+        assert isinstance(deriv, int)
+        if deriv >= len(self.__sample):
+            raise ValueError("""
+The sample on has {len(self.__sample)-1} derivatives.
+Adjust the `n_derivs` parameter of the sampler to access higher derivatives.
+""")
+        return self.__sample[deriv]
+
+
+class CurvePerturbed(sopp.Curve, Curve):
+
+    """A perturbed curve."""
+
+    def __init__(self, curve, sample):
+        r"""
+        Perturb a underlying :mod:`simsopt.geo.curve.Curve` object by drawing a perturbation from a 
+        ``GaussianSampler``.
+
+        Comment:
+        Doing anything involving randomness in a reproducible way requires care.
+        Even more so, when doing things in parallel.
+        Let's say we have a list of :mod:`simsopt.field.curve.Curve`s ``curves`` that represent a stellarator,
+        and now we want to consider ``N`` perturbed stellarators. Let's also say we have multiple MPI ranks.
+        To avoid the same thing happening on the different MPI ranks, we could pick a different seed on each rank.
+        However, then we get different results depending on the number of MPI ranks that we run on. Not ideal.
+        Instead, we should pick a new seed for each :math:`1\le i\le N`. e.g.
+
+        .. code-block::
```

Comment:
I believe `simsopt.geo.curveperturbed` needs to be added to `docs/source/simsopt.geo.rst`. When I tried adding it, and ran sphinx locally, sphinx gave an error about this spot: `/Users/mattland/simsopt/simsopt/src/simsopt/geo/curveperturbed.py:docstring of simsopt.geo.curveperturbed.CurvePerturbed.__init__:13: WARNING: Error in "code-block" directive:
maximum 1 argument(s) allowed, 12 supplied.`